### PR TITLE
Migration to `express-jsdoc-swagger`

### PR DESCRIPTION
**Pull request by JustSamuel**
This PR migrates `express-swagger-generator` to `express-jsdoc-swagger`. This is a large and breaking PR where most of the Jsdoc has to be rewritten. The generation and saving of the swagger file has been implemented.

TODO:
- [ ] Figure out how file uploading works nicely.
- [x] Extract filePattern to function signature.
- [ ] Clean up (remove) Entities Swagger definitions

### Controllers
- All files in `./controller/*.ts` are done.

### Controllers - Request
- All files in `./controller/request/*.ts` are done.

### Controllers - Response
- All files in `./controller/response/*.ts` and `./controller/response/**/*.ts` are  done.

#### Entities
- The usage of Entities swagger definitions has been removed.

### CODE REVIEW

**Comment by JustSamuel** [comment]
### How2Migrate

#### Example of Old Format
```javascript
/**
 * Returns all existing banners
 * @route GET /open/banners
 * @operationId getAllOpenBanners
 * @group banners - Operations of banner controller
 * @param {integer} take.query - How many banners the endpoint should return
 * @param {integer} skip.query - How many banners should be skipped (for pagination)
 * @returns {PaginatedBannerResponse.model} 200 - All existing banners
 * @returns {string} 400 - Validation error
 * @returns {string} 500 - Internal server error
 */
```

#### Changes Required
1. **Description to @summary**
   - Replace the initial description line with `@summary`.
   - Also move the @route to the first position and the summary to the second position.
   - **Find Regex**: `\/\*\*\n\s*\* (.+)\n`
   - **Replace**: `/**\n * @summary $1\n`

2. **@route to CRUD {path}**
   - Replace `@route` with the HTTP method and path.
   - **Find Regex**: `@route (\w+) (.+)`
   - **Replace**: `$1 $2`

3. **@group to @tags**
   - Replace `@operationId` with `@tags`.
   - **Find Regex**: `@operationId (.+)`
   - **Replace**: `@tags $1`

5. **@returns to @return**
   - Replace `@returns` with `@return`.
   - **Find Regex**: `@returns`
   - **Replace**: `@return`

#### Example of New Format
```javascript
/**
 * GET /open/banners
 * @summary Returns all existing banners
 * @operationId getAllOpenBanners
 * @tags banners - Operations of banner controller
 * @param {integer} take.query - How many banners the endpoint should return
 * @param {integer} skip.query - How many banners should be skipped (for pagination)
 * @return {Array<BannerResponse>} 200 - All existing banners
 * @return {string} 400 - Validation error
 * @return {string} 500 - Internal server error
 */
```

#### Additional Notes for Entities
For documenting entities, use `@typedef` to define types. Here's an example:

```javascript
/**
 * @typedef {object} BannerResponse
 * @property {string} name.required - Name/label of the banner
 * @property {string} image - Location of the image
 * @property {number} duration.required - How long the banner should be shown (in seconds)
 * @property {boolean} active.required - Whether the banner is active. Overrides start and end date
 * @property {string} startDate.required - The starting date from which the banner should be shown
 * @property {string} endDate.required - The end date from which the banner should no longer be shown
 */
```

- Replace `{Array.<BannerResponse>}` with `Array<BannerResponse>`.
- Remove `.model` from types.
- For inheritance in typedefs, use `@typedef {allOf|parent} Type`.

**Comment by JustSamuel** [comment]
TODO
Find a fix for integer enums. For the `StripeDepositStatusResponse` I had to change
`* @property {integer} state.required - enum:1,2,3,4 - State of the Stripe deposit. It can be 1 ('CREATED'), 2 ('PROCESSING'), 3 ('SUCCEEDED'), or 4 ('FAILED')`

To 
`@property {integer} state.required - State of the Stripe deposit. It can be 1 ('CREATED'), 2 ('PROCESSING'), 3 ('SUCCEEDED'), or 4 ('FAILED')`
Since express-jsdoc-swagger always creates string enums. 
I opened a bug report [here](https://github.com/BRIKEV/express-jsdoc-swagger/issues/257)

File: src/controller/response/debtor-response.ts [hunk 1]
```diff
@@ -34,46 +34,46 @@ export interface UserToFineResponse {
 }
 
 /**
- * @typedef {BaseResponse} FineResponse
- * @property {DineroObjectResponse.model} amount.required - Fine amount
- * @property {BaseUserResponse.model} user.required - User that got the fine
+ * @typedef {allOf|BaseResponse} FineResponse
+ * @property {DineroObjectResponse} amount.required - Fine amount
+ * @property {BaseUserResponse} user - User that got the fine
```
Comment (Yoronex) [inline src/controller/response/debtor-response.ts:39]: User is not required in JSdoc, but is required in the interface definition.

File: src/controller/response/rbac/action-response.ts [hunk 2]
```diff
@@ -19,11 +19,11 @@
 import RelationResponse from './relation-response';
 
 /**
- * @typedef ActionResponse -
+ * @typedef {object} ActionResponse -
  * The action contains the name of the action and a list of permissions per action.
  * Typically the action name is one of the CRUD values 'create', 'read', 'update', and 'delete'.
  * @property {string} action - The name of the action performed on the entity.
```
Comment (Yoronex) [inline src/controller/response/rbac/action-response.ts:25]: `action` and `relations` are required in the response, but not in the JSdoc.

File: src/controller/response/rbac/entity-response.ts [hunk 3]
```diff
@@ -19,10 +19,10 @@
 import ActionResponse from './action-response';
 
 /**
- * @typedef EntityResponse -
+ * @typedef {object} EntityResponse -
  * The entity contains a name and a list of permissions per action.
  * @property {string} entity - The name of the entity for which the permissions are.
```
Comment (Yoronex) [inline src/controller/response/rbac/entity-response.ts:24]: `entity` and `actions` are required in the response, but not in the JSdoc.

File: src/controller/response/rbac/relation-response.ts [hunk 4]
```diff
@@ -19,12 +19,12 @@
 import { AllowedAttribute } from '../../../rbac/role-manager';
 
 /**
- * @typedef RelationResponse -
+ * @typedef {object} RelationResponse -
  * The relation response contains the name of the ownership relation towards the entity,
  * and the list of attributes for which the role gives access.
  * Typical ownership relations are 'own', 'created', and 'all'.
  * @property {string} relation - The the ownership relation towards the entity.
```
Comment (Yoronex) [inline src/controller/response/rbac/relation-response.ts:26]: `relation` and `attributes` are required in the response, but not in the JSdoc.

File: src/controller/response/rbac/role-response.ts [hunk 5]
```diff
@@ -19,10 +19,10 @@
 import EntityResponse from './entity-response';
 
 /**
- * @typedef RoleResponse -
+ * @typedef {object} RoleResponse -
  * A role contains a unique name, and a list of permissions per entity.
  * @property {string} role.required - The name of the role.
- * @property {Array.<EntityResponse>} entities - The permissions with regards to the entity.
+ * @property {Array<EntityResponse>} entities - The permissions with regards to the entity.
```
Comment (Yoronex) [inline src/controller/response/rbac/role-response.ts:25]: `entities` is required in the response, but not in the JSdoc.

File: src/controller/response/stripe-response.ts [hunk 6]
```diff
@@ -31,20 +31,22 @@ export interface StripePaymentIntentResponse extends BaseResponse {
   clientSecret: string;
 }
 
+// TODO find a fix for integer enums.
```
Comment (Yoronex) [inline src/controller/response/stripe-response.ts:34]: Please add a reference to the bug report you created at the `express-jsdoc-swagger` repository.

File: src/controller/stripe-controller.ts [hunk 7]
```diff
@@ -57,13 +57,13 @@ export default class StripeController extends BaseController {
...
-   * @route POST /stripe/deposit
+   * POST /stripe/deposit
+   * @summary Start the stripe deposit flow
    * @operationId deposit
-   * @group Stripe - Operations of the stripe controller
-   * @param {StripeRequest.model} stripe.body.required - The deposit that should be created
-   * @returns {StripePaymentIntentResponse.model} 200 - Payment Intent information
-   * @returns {string} 500 - Internal server error
+   * @tags Stripe - Operations of the stripe controller
+   * @param {StripeRequest} request.body.requried - The deposit that should be created
```
Comment (Yoronex) [inline src/controller/stripe-controller.ts:64]: `requried` instead of `required`.

File: src/controller/transaction-controller.ts [hunk 8]
```diff
@@ -132,17 +132,17 @@ export default class TransactionController extends BaseController {
...
   /**
-   * Creates a new transaction
-   * @route POST /transactions
+   * POST /transactions
+   * @summary Creates a new transaction
    * @operationId createTransaction
-   * @group transactions - Operations of the transaction controller
-   * @param {TransactionRequest.model} transaction.body.required -
+   * @tags transactions - Operations of the transaction controller
+   * @param {TransactionRequest} request.body.requried -
```
Comment (Yoronex) [inline src/controller/transaction-controller.ts:139]: `requried` instead of `required`.

File: test/unit/controller/transaction-controller.ts [hunk 9]
```diff
@@ -579,12 +579,13 @@ describe('TransactionController', (): void => {
         .get(`/transactions/${trans.id}`)
         .set('Authorization', `Bearer ${ctx.organMemberToken}`);
       expect(res.status).to.equal(200);
-      expect(ctx.specification.validateModel(
+      // TODO Fix the BasePointOfSaleResponse; it always contains a revision and useAuth but this is not reflected in the swagger.
```
Comment (Yoronex) [inline test/unit/controller/transaction-controller.ts:582]: What is this TODO?

File: test/unit/controller/user-controller.ts [hunk 10]
```diff
@@ -1193,7 +1193,7 @@ describe('UserController', (): void => {
           'TransactionReportResponse',
           res.body,
           false,
-          true,
+          false,
```
Comment (Yoronex) [inline test/unit/controller/user-controller.ts:1196]: If I remember correctly, we now allow more data to be returned than just the spec? Why would we suddenly allow this?

File: src/controller/user-controller.ts [hunk 11]
```diff
@@ -384,17 +384,17 @@ export default class UserController extends BaseController {
...
-   * Put an users pin code
-   * @route PUT /users/{id}/authenticator/pin
+   * PUT /users/{id}/authenticator/pin
+   * @summary Put an users pin code
    * @operationId updateUserPin
-   * @group users - Operations of user controller
+   * @tags users - Operations of user controller
    * @param {integer} id.path.required - The id of the user
-   * @param {UpdatePinRequest.model} update.body.required -
+   * @param {UpdatePinRequest} request.body.requried -
```
Comment (Yoronex) [inline src/controller/user-controller.ts:392]: `requried` instead of `required`.

File: src/controller/user-controller.ts [hunk 12]
```diff
@@ -731,16 +731,15 @@ export default class UserController extends BaseController {
 
   // TODO make a specification that can handle undefined attributes.
```
Comment (Yoronex) [inline src/controller/user-controller.ts:732]: I think this TODO is now resolved? If so, please remove it.

**Review by Yoronex — CHANGES_REQUESTED** [review]

File: test/unit/controller/transaction-controller.ts (hunk 9 above)
Comment (JustSamuel) [reply test/unit/controller/transaction-controller.ts:582]: The `product` in the `SubTransactionRowResponse` is a `BaseProductResponse` but also has a `revision` so its neither a `BaseProductResponse` nor a `ProductResponse`.

See line 387 of the `transaction-service.ts`

File: test/unit/controller/user-controller.ts (hunk 10 above)
Comment (JustSamuel) [reply test/unit/controller/user-controller.ts:1196]: Same goes as for the previous remark `TransactionReportResponse` - `data.entries.product` is of type `BaseProductResponse` but also has a `revision` property.

File: test/unit/controller/transaction-controller.ts (hunk 9 above)
Comment (Yoronex) [reply test/unit/controller/transaction-controller.ts:584]: If you open an issue for this (i.e. removing revision from the responses to make it transparent for endusers), I am fine with this. Please refer to the issue on this line.

File: test/unit/controller/user-controller.ts (hunk 10 above)
Comment (Yoronex) [reply test/unit/controller/user-controller.ts:1196]: Same as above. Please add a comment that elaborates why this test is different than all the others (and include the Github issue reference).

**Review by Yoronex — APPROVED** [review]

**Pull request merged**